src/ngraph/evaluator_tensor.hpp

```diff
@@ -29,26 +29,85 @@ namespace ngraph
+    using EvaluatorTensorPtr = std::shared_ptr<EvaluatorTensor>;
+    using EvaluatorTensorVector = std::vector<EvaluatorTensorPtr>;
+
+    NGRAPH_API
+    std::string node_evaluation_failure_loc_string(const Node* node);
+
+    class NGRAPH_API NodeEvaluationFailure : public CheckFailure
+    {
+    public:
+        NodeEvaluationFailure(const CheckLocInfo& check_loc_info,
+                              const Node* node,
+                              const std::string& explanation)
+            : CheckFailure(check_loc_info, node_evaluation_failure_loc_string(node), explanation)
+        {
+        }
+    };
+
+#define NODE_EVALUATION_CHECK(node, ...)                                                           \
+    NGRAPH_CHECK_HELPER(::ngraph::NodeValidationFailure, (node), __VA_ARGS__)
 
     /// \brief A generic handle to (potential) storage with element type and shape information
     class NGRAPH_API EvaluatorTensor
     {
     protected:
-        EvaluatorTensor(const element::Type& element_type, const PartialShape& partial_shape);
+        EvaluatorTensor(const element::Type& element_type,
+                        const PartialShape& partial_shape,
+                        bool is_allocated);
         /// \brief Get type/shape from value
-        EvaluatorTensor(const Output<Node>& value);
+        EvaluatorTensor(const Output<Node>& value, bool is_allocated);
+
+        /// \brief Marks the tensor as allocated
+        void set_is_allocated();
 
     public:
         virtual ~EvaluatorTensor();
-        /// \return pointer to storage
+        /// \return pointer to storage. The element type and shape must be static when this is
+        /// called.
         virtual void* get_data_ptr() = 0;
-
+        /// \return The current element type
         const element::Type& get_element_type() const;
+        /// \return The current partial shape
         const PartialShape& get_partial_shape() const;
+        /// \return the static shape
         const Shape get_shape() const;
-        size_t get_element_count();
-        size_t get_size_in_bytes();
+        /// \return The number of elements in the shape
+        size_t get_element_count() const;
+        /// \return The number of bytes in the shape
+        size_t get_size_in_bytes() const;
+        /// \return True if storage has been allocated
+        bool get_is_allocated() const;
+        /// \brief Set the element type. Must be compatible with the current element type.
+        /// \param node The node being evaluated, for context
+        /// \param element_type The element type
+        void set_element_type(Node* node, const element::Type& element_type);
+        /// \brief Set the shape for allocation. Must be compatible with the current partial shape.
+        void set_allocation_shape(Node* context, const Shape& shape);
+        /// \brief Set the actual shape of the tensor. Dimension 0 can be less than the allocation
```

Comment:
Why only Dimension 0 can be < allocation but not other dimensions? In the case of NonZero op, the output shape is {input_shape_rank, num_non_zero_elem}. Dim 0 might be dynamic, Dim 1 is most likely less then the allocation shape. How would this work if only Dim 0 can be less?

Reply:
Only dim 0 can be shortened without shuffling elements. If you need to shorten something else, your kernel will need to build a temporary result and then copy.

Reply:
I took out this shortening capability.